agent/local/state.go

```diff
@@ -842,6 +843,7 @@ func (l *State) deleteService(id string) error {
 		// todo(fs): some backoff strategy might be a better solution
 		l.services[id].InSync = true
 		l.logger.Printf("[WARN] agent: Service %q deregistration blocked by ACLs", id)
+		metrics.IncrCounterWithLabels([]string{"consul", "acl", "blocked", "service", "deregistration"}, 1, []metrics.Label{{Name: "id", Value: id}})
```

Comment:
In general this looks pretty useful and I totally agree we should have metrics for ACL deny.

The one part I'm not 100% sure on is the label. In practice I don't think it's going to be a huge deal but for customers with tens of thousands of service instances which have unique ids due to being scheduled rather than just left with their service name these labels might be not very meaningful and a significant explosion in metric cardinality. If it were the service name then it would be somewhat better (at least more meaningful and potentially lower cardinality).

That said, these are being collected from agents which presumably in most setups add unique host name labels already anyway so it's only really increasing the cardinality by the average number of services on a single agent I guess.

The winning fact is that in a well-behaved cluster these metrics should never be produced so the cardinality only increases a lot if many service instances start failing ACL checks. 

Of course that might cause an interesting cascading failure mode where a single ACL misconfiguration not only takes out service registrations across the whole cluster for a bit but blows up the metrics system in the process. But I don't think that's a compelling reason not to add it.

Would this still be useful to you without the service ID labels @pierresouchay ? My thought is you still get a warning signal and can dig into logs on a host to get the specifics rather than have many sparse series? Not against it as it is but it's certainly a grey area how much meta data is stored in time-series metric cardinality.

Reply:
@banks We actually already have the message, we are completely blind without any labels:

* I am tracking for months various services being blocked
* If forces to parse all logs on all servers and does not work well since it is not linked directly to the registration itself (and we have several registrations / min on many hosts)
* We had to parse all logs in order to find the culprits, so basically, the only feature we need is the service being blocked
* I have a similar issue with DCs not existing, but that's another topic

But I am for adding the blacklist of labels using https://github.com/armon/go-metrics/pull/77 and https://github.com/hashicorp/consul/pull/4042 .

I am just waiting for https://github.com/armon/go-metrics/pull/77  to rework https://github.com/hashicorp/consul/pull/4042

Reply:
Interesting, that's good context - thanks.

From what I understand, just the fact you are adding these to the agent side would solve a lot of your problem with "blindness" right? Even without the labels, you'd be able to see which _agent_ was issuing those bad requests from metrics, then have only one log to look in and correlate times to see the details of which service instance is responsible (assuming that info is logged which is a separate issue).

I'm not against the labels as I mentioned and for almost all users it won't be much more to add label per service as it will to add label per agent anyway presumably.

Reply:
DONE